testserver/server.py

```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/python
+
+import threading 
+import socket
+import select
+
+
+def consume_socket_content(sock, chunks=65536, timeout=0.5):
+    content = ""
+    more_to_read = select.select([sock], [], [], timeout)[0]
+
+    while more_to_read:
+        new_content = sock.recv(chunks).decode("utf-8")
+
+        if len(new_content) == 0:
+            more_to_read = False # empty recv means the socket disconnected
+
+        else:
+            content += new_content 
+            # stop reading if no new data is received for a while 
+            more_to_read = select.select([sock], [], [], timeout)[0] 
```

Comment:
I'm nervous about this `select` based approach to draining sockets. It runs the risk of being a bit brittle: slow test machines may run into problems, and fast test machines may be delayed up to half a second in a test.

I presume that the reason we're doing this is to avoid problems with needing the test suite to understand HTTP, which I can understand, but we need to be wary of this in the future and consider coming up with a different approach.